crates/shim/src/asynchronous/task.rs

```diff
@@ -0,0 +1,212 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use log::{debug, info};
+use tokio::sync::{MappedMutexGuard, Mutex, MutexGuard};
+
+use containerd_shim_protos::api::DeleteResponse;
+use containerd_shim_protos::protobuf::well_known_types::Timestamp;
+use containerd_shim_protos::protobuf::SingularPtrField;
+use containerd_shim_protos::shim_async::Task;
+use containerd_shim_protos::ttrpc;
+use containerd_shim_protos::ttrpc::r#async::TtrpcContext;
+
+use crate::api::{
+    CreateTaskRequest, CreateTaskResponse, DeleteRequest, Empty, ExecProcessRequest, KillRequest,
+    ResizePtyRequest, ShutdownRequest, StartRequest, StartResponse, StateRequest, StateResponse,
+    Status, WaitRequest, WaitResponse,
+};
+use crate::asynchronous::container::{Container, ContainerFactory};
+use crate::common::ShutdownFn;
+use crate::util::AsOption;
+use crate::TtrpcResult;
+
+/// TaskService is a Task template struct, it is considered a helper struct,
+/// which has already implemented `Task` trait, so that users can make it the type `T`
+/// parameter of `Service`, and implements their own `ContainerFactory` and `Container`.
+pub struct TaskService<F, C> {
```

Comment:
Can we move this to `runc-shim` to align with sync APIs.
See here https://github.com/containerd/rust-extensions/pull/35 for context.

Reply:
I implemented `Task` trait in `shim` crate with this `TashService` template, and the codes in the implemetation are not runtime specific, the `runc-shim` (or any other shim like `kata-shim`, `gvisor-shim` in the future?) do not need to implement the `Task` trait anymore, they just implements the trait exposed in `shim` crate like `ContainerFactory`, `ProcessLifecycle`. and construct their implementation of `Task` trait with those traits implementions.

Reply:
Actually we have an implementation of kata-shim, if every kata-shim implements the `Task` trait, there will be many duplicate codes.